### PR TITLE
Fix type annotation for table literals

### PR DESCRIPTION
**Pull request by axch**

### CODE REVIEW

File: src/lib/Inference.hs [hunk 1]
```diff
@@ -1812,33 +1813,57 @@ checkExtLabeledRow (Ext types (Just ext)) = do
...
+      staticallyKnownIdx (Abs decls res) = unsafeLiftInterpMCatch (evalDecls decls $ evalExpr $ Atom res) >>= \case
+        (Success (IdxRepVal ix)) -> return $ Just ix
+        _ -> return Nothing
+
+      context = "If attempting to construct a fixed-size table not\n" <>
+                "indexed by 'Fin n' for some static n, this error may\n" <>
+                "indicate there was not enough information to infer\n" <>
+                "a concrete index set; try adding an explicit\n" <>
+                "annotation."
+      {-# INLINE staticallyKnownIdx #-}
```
Comment (apaszke) [inline src/lib/Inference.hs:1873]: Please move this next to the function definition

**Review by apaszke — APPROVED** [review]
Looks good!

File: src/lib/Interpreter.hs [hunk 2]
```diff
@@ -321,6 +334,18 @@ interpIndexSetSize ixTy = liftBuilderInterp $ indexSetSize $ sink ixTy
...
 interpIntToIndex ixTy i = liftBuilderInterp $ intToIndex (sink ixTy) (sink i)
 
+-- A variant of `indices` that accepts an expected number of them and
+-- only tries to construct the set if the expected number matches the
+-- given index set's size.
+indicesLimit :: EnvReader m => Int -> IxType n -> m n (Either Int32 [Atom n])
+indicesLimit sizeReq ixTy = do
+  ~(IdxRepVal size) <- interpIndexSetSize ixTy
+  if size == fromIntegral sizeReq then
+    Right <$> forM [0..size-1] \i -> interpIntToIndex ixTy $ IdxRepVal i
```
Comment (apaszke) [inline src/lib/Interpreter.hs:344]: This `forM` will likely run in a virtualized monad that lifts to interpreter at every iteration. Perhaps we could just run it all in the interpreter monad?

File: src/lib/Inference.hs (hunk 1 above)
Comment (axch) [reply src/lib/Inference.hs:1873]: Done.

File: src/lib/Interpreter.hs (hunk 2 above)
Comment (axch) [reply src/lib/Interpreter.hs:344]: Done.

**Pull request merged**